app/services/preserved_object_handler.rb

```diff
@@ -183,37 +183,60 @@ def create_db_objects(status, validated=false)
       # FIXME: what if there is more than one associated pres_copy?
       if incoming_version > pres_copy.version && pres_copy.version == pres_object.current_version
-        results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, pres_copy.class.name)
+        # fine to append this result, because it's true regardless of whether the transaction succeeds
+        full_results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, pres_copy.class.name)
         update_preserved_copy_version_etc(pres_copy, incoming_version, incoming_size, validated)
-        results.concat(update_status(pres_copy, status)) if status
-        results.concat(update_db_object(pres_copy))
-        results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, pres_object.class.name)
+        db_upd_results.concat(update_status(pres_copy, status)) if status
+        db_upd_results.concat(update_db_object(pres_copy))
+        # fine to append this result, because it's true regardless of whether the transaction succeeds
+        full_results << result_hash(ARG_VERSION_GREATER_THAN_DB_OBJECT, pres_object.class.name)
         pres_object.current_version = incoming_version
-        results.concat(update_db_object(pres_object))
+        db_upd_results.concat(update_db_object(pres_object))
       else
-        results << result_hash(UNEXPECTED_VERSION, 'PreservedCopy')
-        results.concat(version_comparison_results(pres_copy, :version))
-        results.concat(version_comparison_results(pres_object, :current_version))
-        results.concat(update_status(pres_copy, status)) if status
+        # these just add result codes about object state w/o touching DB, so can append immediately to rull result list
+        full_results << result_hash(UNEXPECTED_VERSION, 'PreservedCopy')
+        full_results.concat(version_comparison_results(pres_copy, :version))
+        full_results.concat(version_comparison_results(pres_object, :current_version))
+
+        # update_status and update_db_object both touch the db, so same circumspect handling of results from them
+        db_upd_results.concat(update_status(pres_copy, status)) if status
         update_pc_validation_timestamps(pres_copy) if validated
-        results.concat(update_db_object(pres_copy)) if pres_copy.changed?
+        db_upd_results.concat(update_db_object(pres_copy)) if pres_copy.changed?
       end
     end
-    results.concat(upd_results)
+
+    # ok, now we're out of the woods:  if we're here, the transaction is over.  and if it produced no results
+    # of its own, it completed and committed successfully.  so if there were no error codes produced from the
+    # transaction running and committing, return the update results, otherwise, return the transaction failure code(s).
+    if transaction_results.empty?
+      full_results.concat(db_upd_results)
+    else
+      full_results.concat(transaction_results)
+    end
```

Comment:
again, I'm not convinced all of this makes the code clearer.  it certainly seems to make it more complex.  And longer.

I think some of the intent of this work might be better served with better naming -- being clear about updating ActiveRecord object vs updating the database, and being clear about what the transaction code does (I think it is fairly clear -- i'm just a noob to transactions.)

Reply:
This new code is more complex but seems more readable to me.

Reply:
> again, I'm not convinced all of this makes the code clearer

me neither 😄 

>  it certainly seems to make it more complex. And longer.

i would agree